pkg/k8s/watchers/cilium_endpoint_batch.go

```diff
@@ -0,0 +1,202 @@
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package watchers
+
+import (
+	"sync"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+
+	"github.com/cilium/cilium/pkg/k8s"
+	cilium_v2a1 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2alpha1"
+	"github.com/cilium/cilium/pkg/k8s/informer"
+	"github.com/cilium/cilium/pkg/node"
+	"github.com/cilium/cilium/pkg/option"
+	"github.com/sirupsen/logrus"
+)
+
+func (k *K8sWatcher) ciliumEndpointBatchInit(client *k8s.K8sCiliumClient, asyncControllers *sync.WaitGroup) {
+
+	log.Info("Initializing CEB controller")
+	var once sync.Once
+	_, cebController := informer.NewInformer(
+		cache.NewListWatchFromClient(client.CiliumV2alpha1().RESTClient(),
+			cilium_v2a1.CEBPluralName, v1.NamespaceAll, fields.Everything()),
+		&cilium_v2a1.CiliumEndpointBatch{},
+		0,
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				if ceb := k8s.ObjToCiliumEndpointBatch(obj); ceb != nil {
+					k.endpointBatchAdded(ceb)
+				}
+			},
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				if oldCeb := k8s.ObjToCiliumEndpointBatch(oldObj); oldCeb != nil {
+					if newCeb := k8s.ObjToCiliumEndpointBatch(newObj); newCeb != nil {
+						if oldCeb.DeepEqual(newCeb) {
+							return
+						}
+						k.endpointBatchUpdated(oldCeb, newCeb)
+					}
+				}
+			},
+			DeleteFunc: func(obj interface{}) {
+				if ceb := k8s.ObjToCiliumEndpointBatch(obj); ceb != nil {
+					k.endpointBatchDeleted(ceb)
+				}
+			},
+		},
+		nil,
+	)
+	once.Do(func() {
+		// Signalize that we have put node controller in the wait group
+		// to sync resources.
+		asyncControllers.Done()
+	})
+	k.blockWaitGroupToSyncResources(
+		wait.NeverStop,
+		nil,
+		cebController.HasSynced,
+		k8sAPIGroupCiliumEndpointBatchV2Alpha1,
+	)
+
+	go cebController.Run(wait.NeverStop)
+}
+
+// endpointBatchAdded invoked for newly created CEBs, iterates over CoreCiliumEndpoints
+// packed in ceb, converts coreCEP into types.CEP and call endpointUpdated only for remote node CEPs.
+func (k *K8sWatcher) endpointBatchAdded(ceb *cilium_v2a1.CiliumEndpointBatch) {
+
+	nodeIP := node.GetIPv4().String()
+	if option.Config.EnableIPv6 {
+		nodeIP = node.GetIPv6().String()
+	}
+
+	for i, ep := range ceb.Endpoints {
+		log.WithFields(logrus.Fields{
+			"CebName":          ceb.GetName(),
+			"CoreEndpointName": ep.Name,
+		}).Debug("New CEB added, Calling CoreEndpointUpdate")
+		c := k8s.ConvertCoreCiliumEndpointToTypesCiliumEndpoint(&ceb.Endpoints[i])
+		// Skip processing endpointupdate for local node CEPs
```

Comment:
If the node IP changes between cilium-agent restarts, we will receive all CEPs from the previous cilium instance and we will execute `k.endpointUpdated(nil, c)` for those endpoints. Is this safe? I assume it isn't given that this comment is stating that we are skipping the process of local node CEPs.

Reply:
The primary reason for skipping some `CEP`  is that `localNode` already has the latest and greatest CEP, so we don't want any updates for the `localNode`. However, only on the `remoteNodes` we want to update the `CEP`. 

In the scenario you had mentioned, there is nothing wrong in calling `k.endpointUpdated(nil, c)` for the localNode CEPs, this isn't a destructive operation. Eventually `k.endpointUpdated(nil, c)` would be called for updated CEP.

Reply:
Instead of checking nodeIP, can we check if `CEP.ID` is in local endpointManager's store[1]?

[1] https://github.com/cilium/cilium/blob/master/pkg/endpointmanager/manager.go#L49

Reply:
Added comment.